File: Data Structure/Binary Tree/Medium/LET_no222_Count Complete Tree Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        res, h = 0, self.get_height(root)
        while root:
            if self.get_height(root.right) == h - 1:
                res += (1 << h)
                root = root.right
            else:
                res += (1 << h - 1)
                root = root.left
            h -= 1
        return res
        
    def get_height(self, root):
        h = -1
        while root:
            h += 1
            root = root.left
        return h

File: Data Structure/Binary Tree/Medium/LET_no222_Count Complete Tree Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        int res = 0, h = getHeight(root);
        while (root) {
            if (getHeight(root->right) == h - 1) {
                res += 1 << h;
                root = root->right;
            }
            else {
                res += 1 << (h - 1);
                root = root->left;
            }
            h--;
        }
        return res;
    }
    int getHeight(TreeNode* root) {
        return !root ? -1 : 1 + getHeight(root->left);
    }
};